src/lib/operators/index_scan.cpp

```diff
@@ -187,18 +191,19 @@ RowIDPosList IndexScan::_scan_chunk(const ChunkID chunk_id) {
   }
 
   DebugAssert(_in_table->type() == TableType::Data, "Cannot guarantee single chunk PosList for non-data tables.");
-  matches_out.guarantee_single_chunk();
 
-  const auto current_matches_size = matches_out.size();
-  const auto final_matches_size = current_matches_size + static_cast<size_t>(std::distance(range_begin, range_end));
-  matches_out.resize(final_matches_size);
+  auto& chunk_offsets = singleChunkPosList->get_offsets();
+  const auto previous_matches_size = chunk_offsets.size();
+  const auto matches_size = previous_matches_size + static_cast<size_t>(std::distance(range_begin, range_end));
+  chunk_offsets.resize(matches_size);
 
-  for (auto matches_position = current_matches_size; matches_position < final_matches_size; ++matches_position) {
-    matches_out[matches_position] = RowID{chunk_id, *range_begin};
+  for (auto matches_position = previous_matches_size; matches_position < matches_size; ++matches_position) {
+    chunk_offsets[matches_position] = *range_begin;
```

Comment:
That way, we will not see much of the performance gains. What about the idea of pointing into index' offsets as we discussed? This would be a less general type of PosList but a much faster one.

Reply:
Worth benchmarking before we dive into the discussion, but thinking about making the PosList dependent on `AbstractIndex::Iterator` gives me the shivers. :/

Reply:
I just did a test with the current version singleChunkPosList (that is only saving the Row-offset, but does not point into the index) with an index on l_shipdate for the TPCH-6 query. If you look at the [query plan ](https://hyrise-ci.epic-hpi.de/job/hyrise/job/hyrise/job/master/lastStableBuild/artifact/query_plans/tpch/TPC-H_06-LQP.svg), ```l_discount``` might have more effect, but I could not get the indexScan to work with this one. The effects here are really small.

@Bensk1 mentioned an improvement with the PosList pointing into the index of 20%~30% [previously](https://github.com/hyrise/hyrise/pull/2026#issuecomment-595803454), so that might be the better option.
```
+----------------+--------------------+------+--------------------+------+------------+---------+                                                                             │········
| TPC-H 06       | 157.88804626464844 | 9474 | 158.05584716796875 | 9484 | +0%        |  0.6805 |                                                                             │········
| geometric mean |                    |      |                    |      | +0%        |         |                                                                             │········
```

Reply:
> Worth benchmarking before we dive into the discussion, but thinking about making the PosList dependent on `AbstractIndex::Iterator` gives me the shivers. :/

Hm, I am pretty sure that we have discussed that option before. Maybe we did not specifically mention that the PosList has to depend on the AbstractIndex::Iterator though, but certainly that the PosList should somehow "point into the index."

Reply:
I just had the first look at the AbstractIndex interface since a long time. It seems as if all current index implementations actually hold such an `std::vector<ChunkOffset>`. This is only possible because none of the implementations allows for inserts. For now, this means that we could either use `std::vector<ChunkOffset>` or `std::span<ChunkOffset>` if we want to avoid the copy.

For the long term, when considering modifiable indexes, `AbstractIndex::Iterator` seems somewhat reasonable. Note that this requires a virtual method call for both incrementing and dereferencing. We should consider if resolving the PosList type makes sense if we have virtual method calls anyway.

Reply:
To go forward, I would think it is good to follow both approaches: 
1) Make an "Indexscanposlist", that points into the index
2) Make the poslist that only takes chunk_offsets. Here we would start by benchmarking some of the ideas that Markus pointed out [here](https://github.com/hyrise/hyrise/pull/2026#discussion_r399831741)

It might be a good idea to make another merge request for 2), to keep things separated.
We have already tested that 1) can bring a performance boost for TCP-H 6, so I think it is valid to also have an "Indexscanposlist", beside the discussion if a PosList  with offsets (2) is needed or not.

Reply:
Please take into account that each additional PosList increases the compile time. We should only add those that bring a measurable benefit.

Reply:
I just had a discussion with Markus on that. We believe that a second merge request is out of scope for the seminar, it would be too much work. With the new (and amazing) PosList system it should be easy to have a **quick** and dirty benchmark for the two approaches being actually used in an realistic IndexScan, shouldn't it? Thereby, we could verify the results @He3lixxx has seen yesterday and base our decision on which option to implement on that benchmark.

Reply:
I am currently working on benchmarking TPC-H for the first version, to have some data there.

Reply:
Great. I think the TPC-H micro benchmark (that uses the resulting PosList for a second scan) is also sufficient.

Reply:
[Microbenchmarks](https://gist.github.com/XPerianer/47faa74e7933566eb73148d1468f43d3)
In the microBenchmarks, while I get a heavy improvements for the IndexScan, the following tableScan gets a little bit slower, and overall performance does not really increase. This also matches with my observation that I can't get any real benefit in the TPC-H 6 query, (~1% difference).
The code was basically an integration of bensk1/indexscan-poslist in the new master, so I don't understand how you measured the improvements on TPC-H 6.
Maybe I am just missing something here. If you want to look at the code, it's under xperianer/indexscan-poslist (in my fork)

Reply:
I will try running your code on the server were I did the initial measurements. Also, I am surprised that the old style IndexScan is so much faster than the second scan. Before, we never considered the IndexScan to be a realistic option because it was only marginally faster than a TableScan, this is totally different in your microbenchmarks.

Reply:
I added some (useless) benchmarks to your gist. The 30% performance improvement for TPC-H Q6 might have occurred when comparing the new IndexScan to a pure TableScan version. In the past, we could not get any gain from IndexScans for that query.